app/src/test/java/org/dhis2/usecases/main/MainPresenterTest.kt

```diff
@@ -0,0 +1,168 @@
+import org.dhis2.data.schedulers.SchedulerProvider
+import org.dhis2.data.schedulers.TrampolineSchedulerProvider
+import org.dhis2.usescases.login.LoginActivity
+import org.dhis2.usescases.main.MainPresenter
+import org.dhis2.usescases.main.MainView
+import org.hisp.dhis.android.core.D2
+import org.hisp.dhis.android.core.category.CategoryCombo
+import org.hisp.dhis.android.core.category.CategoryOptionCombo
+import org.hisp.dhis.android.core.user.User
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+
+/**
+ * Created by frodriguez on 10/22/2019.
+ *
+ */
+class MainPresenterTest {
+
+    private lateinit var presenter: MainPresenter
+    private val schedulers: SchedulerProvider = TrampolineSchedulerProvider()
+    private val view: MainView = mock()
+    private val d2: D2 = mock()
+    private val preferences: PreferenceProvider = mock()
+    private val workManger: WorkManager = mock()
+
+    @Before
+    fun setUp() {
+        presenter = MainPresenter(view, d2, schedulers, preferences, workManger)
+    }
+
+    @Test
+    fun `Should show username on activity resumed`() {
+        presenterMocks()
+
+        presenter.init()
+
+        verify(view).renderUsername(any())
+    }
+
+    @Test
+    fun `Should save default categoryCombo uid to preferences`() {
+        presenterMocks()
+
+        presenter.init()
+
+        verify(preferences).setValue(DEFAULT_CAT_COMBO, "uid")
+    }
+
+    @Test
+    fun `Should save default categoryOptionCombo uid to preferences`() {
+        presenterMocks()
+
+        presenter.init()
+
+        verify(preferences).setValue(PREF_DEFAULT_CAT_OPTION_COMBO, "uid")
+    }
+
+    @Test
+    fun `Should log out`() {
+        whenever(d2.userModule()) doReturn mock()
+        whenever(d2.userModule().logOut()) doReturn Completable.complete()
+
+        presenter.logOut()
+
+        verify(workManger).cancelAllWork()
+        verify(view).startActivity(LoginActivity::class.java, null, true, true, null)
+    }
+
+    @Test
+    fun `Should block session`() {
+        val pin = "1234"
+
+        presenter.blockSession(pin)
+
+        verify(preferences).setValue(SESSION_LOCKED, true)
+        verify(preferences).setValue(PIN, pin)
+        verify(workManger).cancelAllWork()
+        verify(view).back()
+    }
+
+    @Test
+    fun `Should show filter screen when filter icon is clicked`() {
+        presenter.showFilter()
+
+        verify(view).showHideFilter()
+    }
+
+    @Test
+    fun `Should clear disposable when activity is paused`() {
+        presenter.onDetach()
+
+        val disposableSize = presenter.disposable.size()
+
+        assertTrue(disposableSize == 0)
+    }
+
+    @Test
+    fun `Should open drawer when menu is clicked`() {
+        presenter.onMenuClick()
+
+        verify(view).openDrawer(any())
+    }
+
+    private fun presenterMocks() {
+        // UserModule
+        whenever(d2.userModule()) doReturn mock()
```

Comment:
I know all this mocking is huge. We'll try to return mocks from sdk dagger modules... I'll keep you updated ;)